login mvvm combine

Alexdive opened the pull request.

Alexdive committed:
added combine binding

Alexdive committed:
inject model in init

Alexdive committed:
dynamic height/width

Alexdive committed:
no message

0111b commented on LogIn/LoginViewController.swift:
```diff
@@ -10,32 +10,54 @@ import Combine
 
 class LoginViewController: UIViewController {
```
```suggestion
final class LoginViewController: UIViewController {
```

Alexdive replied:
done

0111b commented on LogIn/UIExtensions.swift:
```diff
@@ -5,23 +5,31 @@
 import UIKit
 
 extension UITextField {
     func loginTextStyle() {
```
maybe something like `applyLoginTextStyle`, because meaning is not clear

Alexdive replied:
done

0111b commented on LogIn/UIExtensions.swift:
```diff
@@ -5,23 +5,31 @@
 
 extension UITextField {
     func loginTextStyle() {
         self.backgroundColor = .systemGray6
```
Also it make sense to make this function private and move to the login view

Alexdive replied:
make sense. Done

0111b commented on LogIn/LoginViewController.swift:
```diff
@@ -10,32 +10,54 @@ import Combine
     
+    var model: LoginViewModelType
+    
+    private let email = PassthroughSubject<String, Never>()
```
To remove boilerplate and copy-paste, its nice to write extension to the TextField with custom publisher.

0111b replied:
Check out https://github.com/CombineCommunity/CombineCocoa/blob/main/Sources/CombineCocoa/Controls/UITextField%2BCombine.swift as an example reference

Alexdive replied:
good one!
Done

0111b commented on LogIn/LoginViewController.swift:
```diff
@@ -45,48 +67,89 @@ class LoginViewController: UIViewController {
+    
+    @objc
+    private func emailTextEntered() {
+        if let text = emailTextField.text {
```
nil checking and other transformations  must be done in the viewModel

Alexdive replied:
done

0111b commented on LogIn/LoginViewModel.swift:
```diff
@@ -0,0 +1,97 @@
+    func transform(input: LoginViewModelInput)
+}
+
+class LoginViewModel: LoginViewModelType {
```
Why it is a `class` and why it is not `final`?

Alexdive replied:
made it final. Class to be able to use @Published wrapper

0111b commented on LogIn/LoginViewModel.swift:
```diff
@@ -0,0 +1,97 @@
+
+class LoginViewModel: LoginViewModelType {
+    
+    @Published private(set) var output: LoginViewModelOutput = .init(emailTint: .systemGray2,
```
You do not need this wrapper here, lets discuss.

0111b commented on LogIn/LoginViewModel.swift:
```diff
@@ -0,0 +1,97 @@
+        var isValid = false
+        
+        input.email
+            .debounce(for: 1, scheduler: RunLoop.main)
```
too long =)

Alexdive replied:
well, its subjective. I don't like when app screaming on me right away

0111b commented on LogIn/LoginViewModel.swift:
```diff
@@ -0,0 +1,97 @@
+        subscriptions.forEach { $0.cancel() }
+        subscriptions.removeAll()
+        
+        var emailTint: UIColor = .systemGray2
```
Lets discuss the flow

Alexdive replied:
adjusted flow

Alexdive committed:
Added Combine extension to UIControl,
UITextfield extension with publisher,
removed unneeded code,
marked classes as final

Alexdive committed:
Adjusted flow

Alexdive committed:
Refactored

Alexdive committed:
renamed view constants to presentation object
